# Python vs Excel: "a comparison"
https://www.youtube.com/watch?v=WbNYbxdW3mk&ab

# Tool set

**I, IMPORT LIBRARY TO WORK WITH DATA**

1, Numpy
- Numpy provides support for large, multi-dimensional arrays and matrices,
along with a variety of high-level mathematical functions to operate on these arrays.
- Common alias for numpy is np
- Syntax:
          import numpy as np

2, Pandas
- Pandas is a popular Python library used for data manipulation and analysis. It provides data structures and functions for working with structured data, such as tabular data, time series data, and more.
- Common alias for pandas is pd
- Syntax:
          import pandas as pd


In [1]:
#Import pandas
import pandas as pd

In [22]:
# Read every rows
pd.set_option("display.max_columns", 28)

**II, UPLOAD FILES USING GOOGLE COLLAB AND READING FILES USING PANDAS**

- syntax:
          df = pd.read_csv('file_name') #turn csv file into dataframe
https://learn.microsoft.com/en-us/power-query/media/combine-files-csv/combine-files-csv-sample-csv-file.png


# Read data - check data file format

In [20]:
# Read data - check data file format
df = pd.read_csv("Cleaned_DS_Jobs.csv")

In [23]:
df.head()

index          Job Title Salary Estimate  \
0      1  Sr Data Scientist        137-171    
1      2     Data Scientist        137-171    
2      3     Data Scientist        137-171    
3      4     Data Scientist        137-171    
4      5     Data Scientist        137-171    

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

         Company Name       Location            Headquarters  \
0         Healthfirst   New York, NY            New York, NY   
1             ManTech  Chantilly, VA             Herndon, VA   
2      Analysis Group     Boston, MA              Boston, MA   
3             INFICON     Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions   New York, NY            New York, NY   

                      Size        Type of ownership  \
0   1001 to 5000 employees   Nonprofit Organization   
1  5001 to 10000 employees         Company - Public   
2   1001 to 5000 employees  Private Practice / Firm   
3    501 to 1000 employees         Company - Public   
4      51 to 200 employees        Company - Private   

                                Industry             Sector  \
0                     Insurance Carriers          Insurance   
1                 Research & Development  Business Services   
2                             Consulting  Business Services   
3  Electrical & Electronic Manufacturing      Manufacturing   
4                Advertising & Marketing  Business Services   

                      Revenue  min_salary  max_salary  avg_salary job_state  \
0    Unknown / Non-Applicable       137.0       171.0       154.0        NY   
1      $1 to $2 billion (USD)       137.0       171.0         NaN        VA   
2  $100 to $500 million (USD)       137.0       171.0         NaN        MA   
3  $100 to $500 million (USD)       137.0       171.0         NaN        MA   
4    Unknown / Non-Applicable       137.0       171.0         NaN        NY   

   same_state  company_age  python  excel  hadoop  spark  aws  tableau  \
0           1           27       0      0       0      0    1        0   
1           0           52       0      0       1      0    0        0   
2           1           39       1      1       0      0    1        0   
3           0           20       1      1       0      0    1        0   
4           1           22       1      1       0      0    0        0   

   big_data        job_simp seniority  
0         0  data scientist    senior  
1         1  data scientist        na  
2         0  data scientist        na  
3         0  data scientist        na  
4         0  data scientist        na

In [ ]:
# Drop unnecessary columns
df.drop(...)

In [37]:
# Drop rows
df1 = df.iloc[:200]

In [35]:
# Rename columns
df1.rename(columns=...)

In [36]:
df1.head(2)

index          Job Title    Salary  \
0      1  Sr Data Scientist  137-171    
1      2     Data Scientist  137-171    

                                     Job Description  Rating Company Name  \
0  Description\n\nThe Senior Data Scientist is re...     3.1  Healthfirst   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2      ManTech   

        Location  Headquarters                     Size  \
0   New York, NY  New York, NY   1001 to 5000 employees   
1  Chantilly, VA   Herndon, VA  5001 to 10000 employees   

        Type of ownership                Industry             Sector  \
0  Nonprofit Organization      Insurance Carriers          Insurance   
1        Company - Public  Research & Development  Business Services   

                    Revenue  min_salary  max_salary  avg_salary job_state  \
0  Unknown / Non-Applicable       137.0       171.0       154.0        NY   
1    $1 to $2 billion (USD)       137.0       171.0         NaN        VA   

   same_state  company_age  python  excel  hadoop  spark  aws  tableau  \
0           1           27       0      0       0      0    1        0   
1           0           52       0      0       1      0    0        0   

   big_data        job_simp seniority  
0         0  data scientist    senior  
1         1  data scientist        na

**III, CHECKING GENERAL INFORMATION**
- Usually we use:
          df.head(number of rows) #display n first few rows
          df.shape                #display total num of rows and cols
          df.describe()           #display summary statistics for the numerical columns in the data
          df.isnull().sum()       #display number of missing values of each column





# Data overview

In [39]:
# Shape

In [25]:
# Info


In [26]:
# Describe


In [41]:
# value counts
df1.value_counts("")           # column name

Type of ownership
Company - Private                 178
Company - Public                   74
Nonprofit Organization             18
Subsidiary or Business Segment     15
Government                          6
Private Practice / Firm             3
-1                                  2
Other Organization                  2
Hospital                            1
Unknown                             1
dtype: int64

# **Data cleaning**

**CHECKING FOR DUPLICATES AND DROPING DUPLICATES**

- syntax:
        
        dup = df[df.duplicated()] #Subsetting duplicated rows

- syntax:
        df = df.drop_duplicates() #Drop duplicated rows

In [8]:
#Checking duplicates


In [9]:
#Dropping duplicates


In [10]:
#checking duplicates again


# **Dealing with missing data**

1, Drop missing values

2, Fill in missing values

a. Mean
                
                mean = df['column'].mean()

-> Replace missing value with mean value
                
                df['column'].fillna(mean, inplace = True)

b. Preceding, following values



In [11]:
#Calculate mean value, fill with mean value


In [12]:
#Fill missing values with bfill


In [14]:
#Fill missing values with ffill


In [15]:
#Checking for missing value again



**CLEANING CATEGORICAL DATA (Dealing with string)**

1, Lowercase string and map to number
- To make change to a column:
                df['column'] = df['column'].(function)


# Data Manipulation

Clean values that don't make sense (outliers)

Select data points with certain characteristics

E.g. Find companies that have a different Headquarter from Location and Rating > 3.5